magma/smart/smart_bits.py

```diff
@@ -0,0 +1,526 @@
+
+    def __str__(self):
+        return str(self._bits)
+
+    def resolve(self, context):
+        self._width_ = len(self._bits)
+        self._signed_ = type(self._bits)._signed
+
+    def eval(self):
+        return self._bits.typed_value()
+
+
+class _SmartBitsMeta(_SmartExprMeta):
+    def __getitem__(cls, key):
+        assert cls is SmartBits
+        signed = False
+        if isinstance(key, tuple):
+            if len(key) == 1:
+                pass
+            elif len(key) == 2:
+                key, signed = key
+            else:
+                raise ValueError(f"{key} unsupported")
+        if _is_int(key):
+            T = Bits[key]
+        else:
+            assert isinstance(key, BitsMeta)
+            T = key
+        name = f"SmartBits[{len(T)}, {signed}]"
+        dct = {"_T": T, "_signed": signed}
+        return type(cls)(name, (cls,), dct)
+
+    def _to_magma_(cls):
+        return cls._T
+
+    def _qualify_magma_(cls, d):
+        return SmartBits[cls._T.qualify(d), cls._signed]
+
+    def _flip_magma_(cls):
+        return SmartBits[cls._T.flip(), cls._signed]
+
+    def _from_magma_value_(cls, value):
+        return cls(value)
+
+    def __repr__(cls):
+        return f"SmartBits[{len(cls._T)}, {cls._signed}]"
+
+
+class SmartBits(_SmartBitsExpr, metaclass=_SmartBitsMeta):
+    def __init__(self, value=None):
+        super().__init__(self)
+        if value is None:
+            value = type(self)._to_magma_()()
+        self._value = bits(value)
+
+    def typed_value(self):
+        if type(self)._signed:
+            return sint(self._value)
+        return uint(self._value)
+
+    def untyped_value(self):
+        return self._value
+
+    def __hash__(self):
+        return hash(self._value)
+
+    def _get_magma_value_(self):
+        return self.untyped_value()
+
+    def __deepcopy__(self, memo):
+        return type(self)(self._value)
+
+    @debug_wire
+    def wire(self, other, debug_info):
+        if not isinstance(other, _SmartExpr):
+            raise ValueError(f"Can not wire {type(self)} to {type(other)}")
+        evaluated, resolved = _eval(self, other)
+        evaluated = evaluated.force_width(len(self))
+        MagmaProtocol.wire(self, evaluated)
+        self._smart_expr_ = resolved  # attach debug info
+
+    def __len__(self):
+        return len(type(self)._T)
+
+    def force_width(self, width):
+        diff = len(self) - width
+        if diff == 0:
+            return copy.deepcopy(self)
+        value = self.typed_value()
+        value = value.ext(-diff) if diff < 0 else value[:-diff]
+        return SmartBits.make(value)
+
+    @staticmethod
+    def make(value):
+        assert isinstance(value, Bits)
+        signed = isinstance(value, SInt)
+        return SmartBits[len(value), signed](value)
+
+    def __str__(self):
+        signed = type(self)._signed
+        return f"SmartBits[{len(self)}, {signed}]({str(self._value)})"
+
+
+class SmartBit(SmartBits[1]):
```

Comment:
Nit: this contradicts the docs which say:

> `SmartBit` which is a literal alias of `SmartBits[1]`

should be 
```Python
SmartBit = SmartBits[1]
``` 
Unless you are some sort of liar

Reply:
😂 good catch. I still need to fix this.

Actually, as you can see even `SmartBit = SmartBits[1]` excludes the possibility of a signed `SmartBit` type (which of course you could do just as `SmartBits[1, True]`.

What about a class `SmartBit` which subclasses and overrides `getitem` to just take in an optional bool for sign? The more I think about it, the more this seems kinda clunky. I'm not sure what you could do besides make a full separate metaclass for `SmartBit`

Reply:
For now, going to go ahead with `SmartBit = SmartBits[1]` literally. Will think about this more. One option is to just get rid of `SmartBit` all together, though I suspect users will want it.